### PR TITLE
[SYCL] Cache devices so they actually compare == when they should.

### PR DESCRIPTION
**Pull request by jbrodman**
Platform device enumeration was creating new device impls every time for the same devices.  This was making it impossible for devices that really should compare equal to compare equal.

Example:
queue q1;
queue q2;

Each q uses the default selector to select the same device - yet they would not compare equal.

Signed-off-by: James Brodman <james.brodman@intel.com>

### CODE REVIEW

File: sycl/source/detail/platform_impl.cpp [hunk 1]
```diff
@@ -15,12 +15,34 @@
...
+#include <map>
 #include <regex>
 
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 namespace detail {
 
+static std::map<RT::PiDevice, std::shared_ptr<device_impl>> device_impl_map;
+static std::mutex device_impl_map_mutex;
+static device host_device;
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.cpp:27]: https://llvm.org/docs/CodingStandards.html#do-not-use-static-constructors

**Comment by jbrodman** [comment]
Note to self: Update this to also fix platforms (same place, same issue).

File: sycl/source/detail/platform_impl.cpp (hunk 1 above)
Comment (romanovvlad) [reply sycl/source/detail/platform_impl.cpp]: Please, avoid using global objects with non-trivial c'tor and d'tor.

Comment (jbrodman) [reply sycl/source/detail/platform_impl.cpp]: I fixed this as best as I could, @romanovvlad

File: sycl/source/detail/device_impl.cpp [hunk 2]
```diff
@@ -251,6 +247,16 @@ bool device_impl::has(aspect Aspect) const {
   }
 }
 
+std::shared_ptr<device_impl> device_impl::getHostDeviceImpl() {
+  static std::shared_ptr<device_impl> HostImpl;
+  static std::once_flag HostImplInit;
+
+  std::call_once(HostImplInit,
```
[the remark this answers is not in the document]
Comment (jbrodman) [reply sycl/source/detail/device_impl.cpp:254]: Was unclear the best way to resolve the "static constructor/destructor" thing.

File: sycl/source/detail/platform_impl.cpp [hunk 3]
```diff
@@ -21,6 +21,57 @@ __SYCL_INLINE_NAMESPACE(cl) {
...
+    HostImpl = std::make_shared<platform_impl>(platform_impl());
+  });
+
+  return HostImpl;
+}
+
+PlatformImplPtr platform_impl::getOrMakePlatformImpl(RT::PiPlatform PiPlatform,
+                                                     const plugin &Plugin) {
+  static std::shared_ptr<std::map<RT::PiPlatform, PlatformImplPtr>> PlatformMap;
+  static std::shared_ptr<std::mutex> PlatformMapMutex;
```
[the remark this answers is not in the document]
Comment (jbrodman) [reply sycl/source/detail/platform_impl.cpp:40]: Probably doesn't.

**Comment by jbrodman** [comment]
I think I've addressed all your points, @romanovvlad. Good suggestions.

**Review by romanovvlad — APPROVED** [review]

**Pull request merged**